run_robot.py

```diff
@@ -0,0 +1,146 @@
+        pico_dir_path = os.path.join(pico_destination_directory, relative_path).replace("\\", "/")
+        temp_dir_path = os.path.join(TEMP_FOLDER, relative_path)
+
+        if not os.path.exists(temp_dir_path):
+            os.makedirs(temp_dir_path)
+        
+        for file_name in files_list:
+            local_file_path = os.path.join(root, file_name)
+            temp_file_path = os.path.join(temp_dir_path, file_name)
+            pico_file_path = os.path.join(pico_dir_path, file_name).replace("\\", "/")
+
+            if not os.path.exists(temp_file_path) or not filecmp.cmp(local_file_path, temp_file_path, shallow=False):
+                # If the file doesn't exist in the temp folder, or it has changed, copy it
+                copy_file_to_pico(files, local_file_path, pico_file_path)
+                # Update the temp folder with the new file
+                shutil.copy2(local_file_path, temp_file_path)
+
+def read_serial_output(port, baudrate=115200, timeout=1):
+    """Read and print serial output from the Pico W."""
+    try:
+        with serial.Serial(port, baudrate, timeout=timeout) as ser:
+            print("Reading serial output...")
+            while True:
+                if ser.in_waiting > 0:
+                    output = ser.read(ser.in_waiting).decode('utf-8')
+                    print(output, end='')  # Print without adding extra newlines
+                time.sleep(0.1)  # Small delay to prevent high CPU usage
+    except Exception as e:
+        print(f"Failed to read serial output. Error: {str(e)}")
+
+def run_pico_script(port, script_name):
+    """Run a MicroPython script on the Pico using ampy."""
+    try:
+        subprocess.run(['ampy', '-p', port, 'run', script_name], check=True)
+        print(f"Successfully ran {script_name} on Pico.")
+    except subprocess.CalledProcessError as e:
+        print(f"Failed to run script {script_name}. Error: {str(e)}")
+
+def list_files_on_pico(port):
+    """List all files and folders on the Pico."""
+    try:
+        pyb = Pyboard(port)
+        files = Files(pyb)
+        file_list = files.ls('/')
+        
+        print("Files and folders on Pico:")
+        for file in file_list:
+            print(file)
+        
+    except Exception as e:
+        print(f"Failed to list files. Error: {str(e)}")
+
+def copy_all_files_to_pico(port, directory=True, main=True, telemetry=True):
+    pyb = Pyboard(port)
+    files = Files(pyb)
+    
+    if directory:
+        compare_and_copy(files, "XRPLib", "lib/XRPLib")
+    if main:
+        copy_file_to_pico(files, "mainprogram.py", "mainprogram.py")
+    if telemetry:
+        copy_file_to_pico(files, "telemetry.txt", "telemetry.txt")
+
+def detect_pico_port():
+    """Auto-detect the Pico W's serial port."""
+    ports = list(serial.tools.list_ports.comports())
+    for port in ports:
+        if "usbmodem" in port.device or "COM" in port.device:  # Adjust this pattern if needed
```

Comment:
This can misidentify other devices as the pico (as happened to me on my desktop). Changing the conditional to `"2E8A:0005" in port.hwid`, to detect picos running MicroPython firmware (as defined [here](https://github.com/raspberrypi/usb-pid)) might help identify the correct serial device.